src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,713 @@
+  mut_encode mut buf
+
+let local_encode idx (local : _ param) buf =
+  let _so, vt = local in
+  write_U32_of_int idx buf;
+  valtype_encode vt buf
+
+let locals_encode typt_f locals buf =
+  let len = List.length locals in
+  let nb_params =
+    match typt_f with
+    | Bt_raw (_ind, (pt, _rt)) -> List.length pt
+    | _ -> assert false
+  in
+  write_U32_of_int len buf;
+  List.iteri (fun i loc -> local_encode (i + nb_params) loc buf) locals
+
+let encode_from_indexed func_local_encode func_import_encode ~is_local data buf
+    =
+  let v = Indexed.get data in
+  match v with
+  | Runtime.Local data ->
+    if is_local then Some (func_local_encode data buf) else None
+  | Runtime.Imported data ->
+    if is_local then None else Some (func_import_encode data buf)
+
+let section_encode id func_encode data buf =
+  let section_buf = Buffer.create 16 in
+  func_encode data section_buf;
+  if Buffer.length section_buf = 0 then ()
+  else (
+    add_char id buf;
+    write_U32_of_int (Buffer.length section_buf) buf;
+    Buffer.add_buffer buf section_buf )
+
+(* type: section 1 *)
+let types_encode func buf =
+  vector_encode_from_namedvalues ~rev_values:true func
+    (fun v buf ->
+      let v = Indexed.get v in
+      match v with
+      | Runtime.Local { type_f = t; locals = _; body = _; id = _ } -> begin
+        match t with
+        | Bt_raw (_ind, (pt, rt)) ->
+          add_char '\x60' buf;
+          paramtype_encode pt buf;
+          resulttype_encode rt buf;
+          Some ()
+        | _ -> assert false
+      end
+      | _ -> None )
+    buf
+
+(* import: section 2 *)
+let imports_encode (func, table, mem, global) buf =
+  let fold_func import_encode_func buf acc data =
+    match
+      encode_from_indexed
+        (fun _ _ -> ())
+        import_encode_func ~is_local:false data buf
+    with
+    | None -> acc
+    | Some () -> acc + 1
+  in
+  let imp_buf = Buffer.create 16 in
+  let { Named.values = func_values; _ } = func in
+  let { Named.values = table_values; _ } = table in
+  let { Named.values = mem_values; _ } = mem in
+  let { Named.values = global_values; _ } = global in
+  let len =
+    List.fold_left
+      (fun acc func ->
+        let func = Indexed.get func in
+        match func with
+        | Runtime.Local _ -> acc
+        | Runtime.Imported { Imported.modul; name; assigned_name = _; desc = _ }
+          ->
+          add_bytes modul imp_buf;
+          add_bytes name imp_buf;
+          add_char '\x00' imp_buf;
+          write_U32_of_int acc imp_buf;
+          (* TODO ? *)
+          acc + 1 )
+      0 func_values
+  in
+  let len =
+    len + List.fold_left (fold_func table_import_encode imp_buf) 0 table_values
+  in
+  let len =
+    len + List.fold_left (fold_func memory_import_encode imp_buf) 0 mem_values
+  in
+  let len =
+    len
+    + List.fold_left (fold_func global_import_encode imp_buf) 0 global_values
+  in
+  write_U32_of_int len buf;
+  Buffer.add_buffer buf imp_buf
```

Comment:
This is really complicated. Here's the beginning of a simplification:

```ocaml
  let imp_buf = Buffer.create 16 in

  let { Named.values = func; _ } = func in
  let { Named.values = table; _ } = table in
  let { Named.values = mem; _ } = mem in
  let { Named.values = global; _ } = global in
  
  let func = keep_imported func in
  List.iteri (fun i { Imported.modul; name; assigned_name = _; desc = _ } ->
    add_bytes modul imp_buf;
    add_bytes name imp_buf;
    add_char '\x00' imp_buf;
    write_U32_of_int acc imp_buf
  ) func;
  
  let fold_func import_encode_func values len =
    let values = List.map (fun v ->
      encode_from_indexed
        (fun _ _ -> ())
        import_encode_func ~is_local:false v imp_buf
    ) values in
    List.filter_map values |> List.length
  in
  
  let len =
    List.length func |>
    fold_func table_import_encode table_values |>
    fold_func memory_import_encode mem_values |>
    fold_func global_import_encode global_values
  in
  
  write_U32_of_int len buf;
  Buffer.add_buffer buf imp_buf
```

It was not tested so it probably does not compile. But I'm sure it can be simplified further by using `keep_local` and `keep_imported` properly instead of using `encode_from_indexed`.

Reply:
confirmed! the code is simpler here too.